### PR TITLE
`DatasetReference` in lookup and `WHERE`

### PR DESCRIPTION
**Pull request by swalker2m**
Rounds out #1050, adding dataset lookup and filtering by `DatasetReference` to the schema.

### CODE REVIEW

File: modules/service/src/main/scala/lucuma/odb/service/DatasetService.scala [hunk 1]
```diff
@@ -30,6 +31,14 @@ import Services.Syntax.*
 
 sealed trait DatasetService[F[_]] {
 
+  /**
+   * Finds the dataset id consistent with the given ids (if any).
+   */
+  def resolveDid(
+    did: Option[Dataset.Id],
+    ref: Option[DatasetReference]
+  ): F[Result[Dataset.Id]]
```
Comment (swalker2m) [inline modules/service/src/main/scala/lucuma/odb/service/DatasetService.scala:40]: I thought I'd need this but then it turned out I didn't.  Still, it seems inevitable and it prompted generalizing the id resolver which is now shared with the `ObservationService` so I decided I'd leave it.

File: modules/service/src/main/scala/lucuma/odb/graphql/mapping/QueryMapping.scala [hunk 2]
```diff
@@ -140,18 +142,31 @@ trait QueryMapping[F[_]] extends Predicates[F] {
...
+    rDid: Result[Option[model.sequence.Dataset.Id]],
+    rRef: Result[Option[DatasetReference]],
+    pred: DatasetPredicates
+  ): Result[Predicate] =
+    (rDid, rRef).parTupled.map { (did, ref) =>
+      and(List(
+        did.map(pred.id.eql).toList,
+        ref.map(r => pred.referenceLabel.eql(r)).toList
+      ).flatten) match {
+        case True => False // neither did nor ref was supplied
```
Comment (toddburnside) [inline modules/service/src/main/scala/lucuma/odb/graphql/mapping/QueryMapping.scala:155]: I like this. 😆

**Review by toddburnside — APPROVED** [review]
LGTM

File: modules/service/src/main/scala/lucuma/odb/graphql/mapping/QueryMapping.scala (hunk 2 above)
Comment (swalker2m) [reply modules/service/src/main/scala/lucuma/odb/graphql/mapping/QueryMapping.scala:155]: Yeah Grackle collapses an empty list down to `Predicate.True` but in this case it shouldn't match anything.

**Pull request merged**